doc/development/06_ca_manager.md

```diff
@@ -0,0 +1,976 @@
+of each CA which tells us about the last successful exchange, or issues, that
+a CA has had when connecting to its parent(s) and repository.
+
+The reason for this split in responsibilities is that tracking the issuance
+of new manifests and CRLs (default: every 16 hours), or the last successful
+check with a parent for resource entitlements (default: every 10 mins) would
+result in an excessive amount of commands and events in the history. In itself
+that is not an issue as long as these historical entries would be valuable
+to rebuilding or understanding the current state of a CA, but the **automatic**
+generation of CRLs and manifests, and historical connections between a CA and
+its parent or repository are not very interesting in this regard.
+
+As a result the `CaManager` is fairly complex because it needs to handle
+these three components in parallel, together with a `CaLocks` structure
+which is used lock read and write access to individual CAs that it manages.
+
+And, for completeness, this component has access to a `KrillSigner` that
+can be used to sign things, and the central `Config` that contains runtime
+settings such as the frequency by which to republish new manifests and CRLs.
+
+Combined we then get this structure:
+
+```rust
+#[derive(Clone)]
+pub struct CaManager {
+    ca_store: Arc<AggregateStore<CertAuth>>,
+    ca_objects_store: Arc<CaObjectsStore>,
+    status_store: Arc<Mutex<StatusStore>>,
+    locks: Arc<CaLocks>,
+    config: Arc<Config>,
+    signer: Arc<KrillSigner>,
+}
+```
+
+Initialization
+--------------
+
+The `CaManager` is instantiated when Krill starts. All its components ultimately
+rely on the `KeyValueStore` - which currently only supports a disk based back-end,
+but which can be modified to support other storage options in future.
+
+Let's first have a look at the initialization code and the included comments, and
+then explain a bit more below.
+
+```rust
+impl CaManager {
+    /// Builds a new CaServer. Will return an error if the CA store cannot be initialized.
+    pub async fn build(config: Arc<Config>, mq: Arc<MessageQueue>, signer: Arc<KrillSigner>) -> KrillResult<Self> {
+        // Create the AggregateStore for the event-sourced `CertAuth` structures that handle
+        // most CA functions.
+        let mut ca_store = AggregateStore::<CertAuth>::disk(&config.data_dir, CASERVER_DIR)?;
+
+        if config.always_recover_data {
+            // If the user chose to 'always recover data' then do so.
+            // This is slow, but it will ensure that all commands and events are accounted for,
+            // and there are no incomplete changes where some but not all files for a change were
+            // written to disk.
+            ca_store.recover()?;
+        } else if let Err(e) = ca_store.warm() {
+            // Otherwise we just tried to 'warm' the cache. This serves two purposes:
+            // 1. this ensures that all `CertAuth` structs are available in memory
```

Comment:
Is it worth saying something about how much memory we are talking about here? I.e. if it's a lot of memory what would happen on a fairly small server, does Krill just crash out and thus has a minmum amount of RAM it needs, or is this stuff small and it's just about keeping it closer to the CPU for faster access (and if that is so, why is fast access important, is it really slow without and has a noticeable impact or is this just an nice-to-have optimisation?) ? (apologies if this is addressed further down this document)

Reply:
This is an optimisation, because if we don't keep the structs in the cache, then we would always have to deserialize snapshots or events when we access them. Now we just have the overhead of serializing when saving, but we don't deserialize until we do a restart.

I think the memory usage is very unlikely to be an issue. But to be fair we have not done measurements on systems with many CAs in a single Krill instance. Also, to be fair, the CPU overhead is not likely to be a real issue either - so deserializing on demand may not be problematic at all..

So, we should probably do some measurements in future. We will also need to review the caching when we want to support multiple active nodes.

Still, cache or no cache, I think it's good to do this warmup as a way to verify that the state can be rebuilt when we start.